### ISSUE TITLE
FlxSprite+FlxAtlas distortions on native targets

### ISSUE THREAD

**Issue by starry-abyss** [HaxeFlixel/flixel#1861]
- **Flixel version:** dev, c561696
- **OpenFL version:** 3.6.1
- **Lime version:** 2.9.1
- **Affected targets:** neko, windows

Hello, I've come across an issue with sprite rendering using atlases. It doesn't occur on Flash, and it doesn't occur when not using FlxAtlas nodes for sprite graphics. It _seems_ to occur more frequently with 3x zoom, and subpixel sprite positions.

Here is a little demo that shows the issue:
https://github.com/starry-abyss/render_test_cpp

Here are some grabs from main project:
http://imgur.com/a/cqvbM

Take attention to bear's eyes, ears and honey cut into halves with diagonal corner-to-corner line (apparently, triangles in quad). Also on GIFs you can see that the frame with bear's eyes shut is rendered differently from other frames.

**Comment by Beeblerox** [comment]
i'm sorry. i don't quite understand what is the problem. Could you describe it in another words?

**Comment by starry-abyss** [comment]
Sure.
The problem is that pixels are drawn imprecisely in certain scenarios (atlas is 100% relevant): 
![distortion](http://i.imgur.com/SAIsa7d.png)

Also, the demo above is a bit interactive (press Space to switch between modes).
Note that you **won't see** the distortion **on Flash** target

**Comment by Beeblerox** [comment]
I think that this is a general problem, which could be seen not only flixel but in any other opengl-based renderer. So possible solution could be round rendering position of the sprite.
Try to add `FlxG.camera.pixelPerfectRender = true;` somewhere in your state's `create()` method. This will force rounding position of all objects, being rendered on default camera.

**Comment by starry-abyss** [comment]
Thanks for the answer.
With ```FlxG.camera.pixelPerfectRender=true``` everything looks jittery.
And I don't think this is general, because exactly the same input, but without FlxAtlas, gives pixel perfect results (please check demo when you can). For me it looks like atlas is adding to imprecision.

**Comment by Beeblerox** [comment]
Will try to do it tonight

**Comment by Beeblerox** [comment]
i'm afraid it's something with opengl renderer (not on flixel level, but on openfl level). 
i will create simple demo without flixel dependency to proof it. 
@starry-abyss will you help me with checking demo when it will be ready?

**Comment by starry-abyss** [comment]
@Beeblerox Thanks for your effort, yes, you can reach me!

**Comment by larsiusprime** [comment]
Hey there guys!

I've had this issue before (or at least one that looks like it), and here's my observations:

- It only happens in certain resolutions
- It only happens on certain people's machines (suggesting a driver-related issue)
- Aggressively rounding/flooring object positions to the integer level *usually* fixes it

**Comment by starry-abyss** [comment]
@larsiusprime Hello, regarding your points:
1. Yes, it seems to happen more frequent when I use 3x zoom, rather than for example 2x or 4x
2. It was actually an artist who first noticed it, and he is on Ubuntu + neko. I forgot to mention it. But he uses GTX660 video card too, if it matters
3. Yes, it helps, but it decreases smoothness a lot (camera scroll, sprite tween)

**Comment by larsiusprime** [comment]
Hopefully we can discover the underlying cause, because I agree that manually rounding your sprite positions seems like it should be unnecessary.

That said, if you're stuck and you need a work-around -- perhaps you could add a video option for aggressive rounding as a "compatibility mode" for people with weird drivers?

**Comment by Tiago-Ling** [comment]
If you aren't doing this already, check if you can keep the movement /
camera smoothness by using `pixelPerfectRender = true` and
`pixelPerfectPosition = false`.

I think in your case what matters is flooring / rounding when rendering.

Tiago Ling Alexandre
Tel: +55 41 8819-3191


2016-05-31 11:05 GMT-03:00 Lars Doucet <notifications@github.com>:

**Comment by starry-abyss** [comment]
@larsiusprime It's not that urgent yet actually, and my main workaround is not using atlas - like OFF mode in my demo above (which can affect me though on Android performance, maybe your suggestion will be better then).
@Tiago-Ling pixelPerfectRender and pixelPerfectPosition are introducing shaking of sprites, so unfortunately. I made a GIF, it's less noticable when recorded though:
![Imgur](http://i.imgur.com/Ix61HBQ.gif)

**Comment by Tiago-Ling** [comment]
Yeah, i had this problem in my last project and spent a long time
investigating it. What i found were workarounds. These are the settings
that made a difference in the end for my sprites when scrolling the camera:

   - I didn't use pixelPerfectRendering nor pixelPerfectPosition (both set
   to false because of the same problem your gif has shown - i just suggested
   it because it could end up producing different results)
   - I used antialiasing = true
   - I used TexturePacker to create all my atlases - one thing that solved
   some of the artifact problems, especially at the frame borders was to set
   the property "extrude" to 1.



Tiago Ling Alexandre
Tel: +55 41 8819-3191


2016-05-31 11:43 GMT-03:00 Igor S. <notifications@github.com>:

**Comment by starry-abyss** [comment]
@Tiago-Ling Thanks, I'll keep it in mind!
My distant-future plan is using the game itself to bake atlas for next runs :-)

**Comment by starry-abyss** [comment]
@Beeblerox, any updates on this?

**Comment by Beeblerox** [comment]
i haven't had time for this, sorry
will start working on example tonight.

**Comment by Beeblerox** [comment]
ok, sorry for the delay, but here what i got.
i've managed to create tilesheet version of the test (without flixel dependency): https://yadi.sk/d/crGC0KWysRPJF
Plus my default resolution is 1366x768, so it can't run your test at 1200x900 and the actual stage size on my machine is 1200x768, which means that game scale factor isn't 3.0 but 2.473 - this makes pixel distortions even more noticeble. 
so i've added ability to switch test scale by pressing up and down buttons (just 2 values of 3 and 2.473).
Can you compile and run this test in different modes (with atlas and without, with scale of 3 and 2.473) and tell me which results do you get?

**Comment by Beeblerox** [comment]
But there is concern: Joshua removed Tilesheet from the dev version of openfl and closed all related issues.

**Comment by starry-abyss** [comment]
@Beeblerox I appreciate your work!
What I see when I run your demo on **neko** or **windows**:
1) In 2.473x mode I see zooming artifacts, but atlas mode looks the same as non-atlas
2) In 3x mode atlas version has distortions (seem identical to my demo), non-atlas doesn't

So what can be done in this regard? Fork Tilesheet into HaxeFlixel? Do tests with OpenFl Tilemap?

**Comment by Beeblerox** [comment]
Yes, i want to make Tilemap version of the test, but will need dev version of openfl (since only dev version supports tile scaling).
And some sort of Tilesheet fork is on my list also.

**Comment by Beeblerox** [comment]
will make Tilemap test tonight

**Comment by starry-abyss** [comment]
Ok. Will I need -Dnext to run with openfl-dev Tilemap?

**Comment by Beeblerox** [comment]
don't think so. there is no legacy mode in dev version of openfl, as far as i know.

**Comment by Beeblerox** [comment]
Here is Tilemap version of the test: https://yadi.sk/d/WO-EfVBAsRVm9
But i had to hack in GLTilemap class to make easy tile animations (modified file included in archive, and can be found at "source/hack" folder. you'll have to replace this file manually if you want to see working animations). i'll make PR to openfl with similar changes.
Here is compiled version of the test: https://yadi.sk/d/MeaXa2HwsRVnv
It looks the same as Tilesheet version for me. But i wonder if it is for you as well.

**Comment by starry-abyss** [comment]
From source on windows target and your binary provide this in atlas in 3x: http://imgur.com/F6He0fQ. The honey is OK, only bear's ear area is distorted. But I think maybe objects coordinates are different in this test?

2.473x is similar atlas and non-atlas, and 3x non-atlas is sharp.
And on neko target I get a crash (not sure if this log says something useful):
```
Module.hx:207: onGamepadConnect (module)
AL lib: (EE) alc_cleanup: 1 device not closed
```

**Comment by Beeblerox** [comment]
i haven't modified coordinates (they should be the same).
Will try to play with Tilemap class sources tomorrow, maybe i'll get something useful.

**Comment by starry-abyss** [comment]
Second take.
All objects in atlas mode appear to be higher (as in position.y), except if bear looks left only his left-bottom half is higher (distortion). All bears' legs in **non-atlas** mode are shorter than they should be: http://imgur.com/keU68JU.

Going to look through previous demos to compare.

**Comment by starry-abyss** [comment]
1) Tilesheet demo: http://imgur.com/bsbSxO0 and http://imgur.com/UNodD32
In non-atlas bears' legs are shorter. And honey's pixels are often broken too: ![Imgur](http://i.imgur.com/J7AcOzz.png).
In atlas if bear looks **right** you have about 50/50 chance: a) only his left-bottom half is higher (=shifted up); or b) whole bear is higher. Same for honey, either whole is higher or half. If bear looks left, his left-top half is higher, and his right bottom half is higher and shifted right too.

2) My demo: http://imgur.com/TNsobu2 and http://imgur.com/wxC4Ewm
Seems the same as Tilesheet

**Comment by starry-abyss** [comment]
@Beeblerox Happy Independence Day!
I've ported the demo to SDL2 built-in hardware renderer (independent of Haxe :-) ). Here's a screenshot: http://imgur.com/39KXYsE. This is how I expect it to render in HaxeFlixel and OpenFl. Note that SDL2 doesn't use float coordinates to render (all cast to int). Also there is no hotkey yet to turn off atlas.
Going to port it to SDL2+OpenGL to do some experiments.

**Comment by Beeblerox** [comment]
@starry-abyss Thank you!
Yeah, i understand that it should look like that. I'll try to modify Tilesheet and TIlemap shaders to use integer coordinates too. Will share results later. Ok?

**Comment by starry-abyss** [comment]
Sure. I'll be waiting for the news from you!

**Comment by Beeblerox** [comment]
ok, i've tried couple solutions found on the internet:
1) floor vertex coordinates
2) add half texel offset to uv coordinates

both solutions seems to work only when scale factor is integer number (3.0), but don't work when scale isn't integer (2.473).
Don't know where to go now actually. Will try to search alternative solution.

Do you want me to upload compiled projects with both workarounds to run on your machine?

**Comment by starry-abyss** [comment]
Maybe some of the solutions would also fix the infamous tilemap tearing?

Only integer is OK, if 2.473 doesn't look worse than earlier :-)

At the moment, I'm not at home, but if testing is needed, I could check the compiled version (at another PC).

**Comment by Beeblerox** [comment]
ok. i'll upload them tonight or tomorrow morning then (need to recompile everything again)

**Comment by Beeblerox** [comment]
Here are compiled demos with some variations:
1. Tilesheet demo with half texel uv offset: https://yadi.sk/d/p-CRv_v3sSQuX
2. Tilesheet demo with vertex coordinates floor: https://yadi.sk/d/eYJD0MvTsSQvr
3. Tilemap demo with half texel uv offset: https://yadi.sk/d/2XzVUvf2sSQxJ
3. Tilemap demo with vertex coordinates floor: https://yadi.sk/d/2dCZ7ziEsSQyw

Tilesheet and Tilemap versions of the same solution (half texel offset or vertex flooring) should look the  same, i think.

About tilemap tearing issue: there is solution which involves tileset modification (which could be done at runtime). i'll provide sample tomorrow. And i'll test sugested methods with tilemap tearing as well.

**Comment by starry-abyss** [comment]
@Beeblerox Great, looks like the real progress! 1, 2, and 4 look perfect to, seems we need more complex tests now!
Tilemap_half_texel_offset crashes at start (maybe you built it in release mode? I have the crashes with -Dnext in release, which I didn't have time to fix, @larsiusprime gave possible workaround with ANGLE and Lime rebuilding, but I had compile errors, and didn't touch it since then).

As for tilemap tearing, I already know that workaround, but it's not a real solution, isn't it? ;-)

Any chance getting Tilesheet version(s) into HaxeFlixel to test in my main project? (I don't need sprite rotation and scaling yet)

**Comment by starry-abyss** [comment]
I also noticed that only one of the 3 working demos in 2.473 gives me similar looking bear's eyes: Tilesheet_vertex_floor. Other two distort eyes differently when sprites are in different positions (but it is the case in current HaxeFlixel too, as far as I remember).

**Comment by Beeblerox** [comment]
I'll try to present something today.
Btw, what openfl and lime versions do you use?

**Comment by starry-abyss** [comment]
Good, lime 2.9.1, openfl 3.6.1 (both unmodified from haxelib). When I tested Tilemap, I used develop branches of them + your hack.

**Comment by Beeblerox** [comment]
I've tested these 2 solutions with tilemaps and can say following:
1. Half texel offset solution makes it even worse.
2. Vertex coordinates flooring seems to kinda work, but i need to hear your opinion on that matter.

In order to use coordinates flooring you'll need to replace one line:
```
'   gl_Position = vec4((${Uniform.ProjectionMatrix} * vec3(${Attrib.Position}, 1.0)).xy, 0.0, 1.0);',
```
with the:
```
'	gl_Position = vec4((${Uniform.ProjectionMatrix} * vec3(floor(${Attrib.Position}.x), floor(${Attrib.Position}.y), 1.0)).xy, 0.0, 1.0);',
```
in openfl's 3.6.1 `openfl._internal.renderer.opengl.shaders2.DefaultShader`

**Comment by starry-abyss** [comment]
Looks promising, I think, you can call floor() on the whole ${Attrib.Position} at once.
Going to test it

**Comment by Beeblerox** [comment]
yeah, you're right about flooring all vector coordinates at once. i'm not expirienced with shaders

**Comment by starry-abyss** [comment]
How to force picking up the new shader (windows target)? Do I need to rebuild openfl (thought it was building when you build the game)?

**Comment by Beeblerox** [comment]
you just need to replace one line in DefaultShader

**Comment by starry-abyss** [comment]
I don't see any difference, even tried to return constant red color from fragment shaders. Maybe they are not used in openfl legacy mode?

**Comment by Beeblerox** [comment]
oh, forgot to tell you about that. DefaultShader is used in openfl's next mode only.
in order to enable it you need to add `<set name="next"/>` line before `<haxelib name="flixel" />` in your project.xml

**Comment by Beeblerox** [comment]
in order to make it work in legacy mode you'll need to modify lime, which i don't want to do.

**Comment by starry-abyss** [comment]
I have to confess that 'next' in HaxeFlixel doesn't work out-of-the-box for me, I have different issues with different projects, mostly crashes.
Hence I can't test right now in the project where most of artifacts occur.
Is there an upgrade guide?

As for the demo I posted here:
In flixel-dev in debug keyboard input is broken, and in release I get this: [crash](http://i.imgur.com/UwBYCKW.png).

**Comment by Gama11** [comment]
There is no such thing as an upgrade guide, it should mostly just work out of the box.

That is a known issue: #1827

**Comment by Tiago-Ling** [comment]
Best way to test next is to compile from command line using "-Dnext" flag.
Simply putting <set name="next" /> doesn't seem to work for me, but doing
"lime test windows -Dnext" does!

Tiago Ling Alexandre
Tel: +55 41 8819-3191


2016-06-13 8:57 GMT-03:00 Igor S. <notifications@github.com>:

**Comment by Gama11** [comment]
`<set name="next" />` works just fine, as long as you place it before `<haxelib name="flixel" />`.

**Comment by starry-abyss** [comment]
Good news is the demo picked up new shader now, and seems to render OK in both atlas and non-atlas.
I probably need some movement and tilemaps in there now

**Comment by starry-abyss** [comment]
'next' works out of the box, if I use Flash, which I mostly don't use

**Comment by Gama11** [comment]
There is no "next" on Flash or HTML5. :D

**Comment by Tiago-Ling** [comment]
Should work out of the box for most cpp stuff, at least in my experience.
What is not working in next for you?

Tiago Ling Alexandre
Tel: +55 41 8819-3191


2016-06-13 9:16 GMT-03:00 Gama11 <notifications@github.com>:

**Comment by starry-abyss** [comment]
@Gama11, such a sly way to make it work out of the box :-P
@Tiago-Ling, started to collect issues to reply to you, and after some builds the main project started working =-O It must be some compilation cache issues?!
Tilemap tearing seems to be gone (at least in one known spot), problem discussed here too. Reverted shader - problems returned

**Comment by Tiago-Ling** [comment]
Yeah, it is always good to make a clean build when changing between legacy
and next. Only problematic thing i found so far in next is screen size /
scale modes in Android (didn't test iOS): inconsistent sizes across devices
with different resolutions.

Tiago Ling Alexandre
Tel: +55 41 8819-3191


2016-06-13 9:51 GMT-03:00 Igor S. <notifications@github.com>:

**Comment by starry-abyss** [comment]
@Tiago-Ling, to be honest, I made clean build from the start (in HaxeDevelop sense of clean build), but let's sum up:
1) Not much of an issue really:
```
	#if openfl_legacy
			var bytes = ByteArray.readFile(fullPath);
			var bitmapData = BitmapData.loadFromBytes(bytes);
	#else
			var bytes = File.getBytes(fullPath);
			var bitmapData = BitmapData.fromBytes(bytes);
	#end
```

2) Crash at start like (neko):
[Imgur](http://i.imgur.com/ti5AmEk.png)

and like (windows):
[Imgur](http://i.imgur.com/eo19fgb.png)

3) Instead of segfault I can get the message like (windows): [Imgur](http://i.imgur.com/9RcHiy3.png)
it seems like some reflection stuff, like tweening

4) Debug mode keyboard input
5) FlxLineShape was gray, became semi-transparent gray; FlxSprite was semi-transparent purple, became black

**Comment by starry-abyss** [comment]
@Beeblerox The movement is as smooth in main project as with original setup (in legacy). I have some strange thing though, but it was always the case, I think, I should study it separately:
![Imgur](http://i.imgur.com/4oq4tuV.png)

**Comment by Beeblerox** [comment]
@starry-abyss what is the problem do you have?

**Comment by starry-abyss** [comment]
@Beeblerox I think when scaling the game, all objects' pixels should be aligned as if I rendered them without scaling and then upscaled the whole frame

**Comment by Beeblerox** [comment]
Ah. i get it. There is pull request for that https://github.com/HaxeFlixel/flixel/pull/1826
this is the scale mode which does what you want. It works on openfl next only. but it hadn't been merged yet. i don't know why. i guess we need to ask @Gama11

**Comment by Beeblerox** [comment]
Plus this scale mode will be broken with dev version of openfl, which makes me sad. 
Joshua is breaking so much cool stuff.

**Comment by starry-abyss** [comment]
Isn't there way to floor vertices taking zoom into account? Like floor(pos / zoom) * zoom?
Also in future could be existing code just used on top of some custom tilesheet with the same interface as old Tilesheet?

**Comment by starry-abyss** [comment]
Yes, I understand that method from PR is the most correct one

**Comment by Beeblerox** [comment]
i think i can (must) create replacement for Tilesheet class in near future, since it had been dropped from development version of openfl

**Comment by Beeblerox** [comment]
i'll put this issue on hold. and will work on Tilesheet replacement then

**Comment by starry-abyss** [comment]
@Beeblerox OK!

**Comment by Beeblerox** [comment]
But i can't promise that i'll make it very soon

**Comment by starry-abyss** [comment]
My problems with 'next' seem to be tied to audio loading and playback, going to investigate it
